<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Балансирование-классов" data-toc-modified-id="Балансирование-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Балансирование классов</a></span></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки</a></span></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li><li><span><a href="#Эксперименты" data-toc-modified-id="Эксперименты-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Эксперименты</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Балансирование-классов-параметром-class_weight='balanced'" data-toc-modified-id="Балансирование-классов-параметром-class_weight='balanced'-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Балансирование классов параметром class_weight='balanced'</a></span></li><li><span><a href="#Увеличение-выборки-на-множитель-4" data-toc-modified-id="Увеличение-выборки-на-множитель-4-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Увеличение выборки на множитель 4</a></span></li><li><span><a href="#Уменьшение-выборки-на-множитель-0.25" data-toc-modified-id="Уменьшение-выборки-на-множитель-0.25-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Уменьшение выборки на множитель 0.25</a></span></li><li><span><a href="#Увеличение-выборки-на-множитель-4-и-параметр-class_weight='balanced'" data-toc-modified-id="Увеличение-выборки-на-множитель-4-и-параметр-class_weight='balanced'-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Увеличение выборки на множитель 4 и параметр class_weight='balanced'</a></span></li><li><span><a href="#Подобранный-оптимальный-множитель-для-увеличения-выборки" data-toc-modified-id="Подобранный-оптимальный-множитель-для-увеличения-выборки-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Подобранный оптимальный множитель для увеличения выборки</a></span></li></ul></li><li><span><a href="#Общий-вывод-по-работе" data-toc-modified-id="Общий-вывод-по-работе-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод по работе</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

<b>Что-то вроде дисклеймсера:</b>

В этой раблоте значения большинства гиперпараметров подбираются в циклах, когда модель раз за разом обучается с новыми гиперпараметрами. От этого тетрадь достаточно долго подсчитывается. Прошу понять и простить. :)

## Подготовка данных
<sup>[Перейти к содержанию](#Содержание)</sup>

In [1]:
#библиотеки
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#всякое
from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle
#модели
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
#метрики
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

rs = 12345

data = pd.read_csv('/datasets/Churn.csv')

Для начала посмотрим на сами данные.

In [2]:
data.info()
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             9091 non-null float64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

В представленных данных зачем-то есть признак, дублирующий индексы. При построении модели он определенно не нужен и будет мешать, так как модель будет пытаться использовать его для предсказания. Поэтому удалим его. 

In [3]:
data = data.drop(axis=1, columns='RowNumber')

В данных у нас есть ID каждого конкретного клиента и его фамилия. Вероятность ухода клиента из банка не может зависеть от индивидуального номера или от фамилии. Так что эти признаки тоже необходимо будет исключить, чтобы модель не пыталась строить на них свои предположения.

In [4]:
data = data.drop(axis=1, columns=['CustomerId', 'Surname'])

Кроме того, в датасете представлены два категориальных признака: Geography и Gender. Сначала посомтрим, какие значения принимают эти признаки.

In [5]:
data['Gender'].value_counts()

Male      5457
Female    4543
Name: Gender, dtype: int64

Соотношение мужчин к женщинам 55/45, почти поровну.

In [6]:
data['Geography'].value_counts()

France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64

Французов 50%, немцев и испанцев примерно по 25%. хоть французов больше, нельзя сказать, что данные слишком перекошены.

Категориальные параметры Geography и Gender можно преобразовать в численные с помощью техники OrdinalEncoder. Однако, из-за пропусков в признаке Tenure (сколько лет человек является клиентом банка), эта техника не работает. Проанализируем, сколько пропусков и как они представлены.

In [7]:
data['Tenure'].value_counts(dropna=False)

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
NaN     909
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: Tenure, dtype: int64

Пропусков практически столько же, сколько данных в каждой из групп. Разве что 0 и 10 лет выбиваются. Можно предположить, что часть данных 0 и 10 лет по ошибке выгрузилась пропуском, однако мы не можем знать этого наверняка. Восстановить эти данные нам неоткуда, а этот признак чисто интуитивно может сильно влиять на результат. Так что придется удалить 10% датасета с пропусками.

In [8]:
data = data.dropna()

После чего наконец преобразуем категориальные данные в численные.

In [9]:
encoder = OrdinalEncoder()

encoder.fit(data) 

data = pd.DataFrame(encoder.transform(data), columns=data.columns)

Ну и наконец изучим соотношение значений ключевого признака.

In [10]:
len(data[data['Exited'] == 1])/len(data)

0.2039379606203938

В представленных данных всего 20% значений ключевого признака равны 1. Это может создать проблемы при обучении модели. Далее в проекте мы проанализируем дисбаланс и решим, что будем с ним делать.

## Исследование задачи
<sup>[Перейти к содержанию](#Содержание)</sup>

Для начала, разобьем имеющиеся данные на датасет с целевыми и прочими признаками. А затем разобьем эти датасеты на три части: учебную, валидационную и тестовую. Так как функции для разбивки на три части мне найти не удалось, просто применим разбивку на две части дважды.

Сначала разобьем датасет на две части в 20% и 80% от исходного датасета. Затем разобьем полученные 80% в пропорции 25% и 75%.  
Таким образом мы получим три выборки:

* учебную, составляющую 60% от исходного датасета
* валидационную, 20% от исходного датасета
* тестовую, 20% от исходного датасета

In [11]:
target = data['Exited']
features = data.drop('Exited', axis=1)

features_train_test, features_valid, target_train_test, target_valid = train_test_split(
    features,
    target,
    test_size=.20,
    random_state=rs)

features_train, features_test, target_train, target_test = train_test_split(
    features_train_test,
    target_train_test,
    test_size=.25,
    random_state=rs)

print('Размер обучающей выборки относительно всего датасета {:.1%} '.format(len(features_train)/len(data)))
print('Размер валидационной выборки относительно всего датасета {:.1%} '.format(len(features_valid)/len(data)))
print('Размер тестовой выборки относительно всего датасета {:.1%} '.format(len(features_test)/len(data)))

Размер обучающей выборки относительно всего датасета 60.0% 
Размер валидационной выборки относительно всего датасета 20.0% 
Размер тестовой выборки относительно всего датасета 20.0% 


Для начала обучим на несбалансированной выборке модель древа решений. В цикле подберем оптимальную глубину древа, при которой F1-метрика будет максимальной.

In [12]:
%%time

max_f1 = 0
best_depth_tree = 0
for depth in range(2, 16):
    model_tree = DecisionTreeClassifier(random_state=rs, max_depth=depth)
    model_tree.fit(features_train, target_train)
    predicted_tree_valid = model_tree.predict(features_valid)
    if f1_score(target_valid, predicted_tree_valid) > max_f1:
        max_f1 = f1_score(target_valid, predicted_tree_valid)
        best_depth_tree = depth

print('Оптимальная глубина дерева, дающая максимальное значение метрики F1, равное', max_f1)
print('max_depth =',best_depth_tree)

Оптимальная глубина дерева, дающая максимальное значение метрики F1, равное 0.5580589254766031
max_depth = 5
CPU times: user 234 ms, sys: 589 µs, total: 234 ms
Wall time: 233 ms


Размер F1-метрики ниже требуемого. Нужно исследовать другие модели.

Попробуем модель случайного леса. Сразу переберем гиперпараметры в цикле, с целью подобрать оптимальные.

In [13]:
%%time

max_f1 = 0
best_estimators_forest = 0
best_depth_forest = 0
for estimators in range(1, 51, 5):
    for depth in range(2, 16, 1):
        model_forest = RandomForestClassifier(random_state=rs, n_estimators=estimators, max_depth=depth)
        model_forest.fit(features_train, target_train)
        predicted_forest_valid = model_forest.predict(features_valid)
        if f1_score(target_valid, predicted_forest_valid) > max_f1:
            max_f1 = f1_score(target_valid, predicted_forest_valid)
            best_estimators_forest = estimators
            best_depth_forest = depth
print('Оптимальные параметры, дающие максимальное значение метрики F1, равное', max_f1)
print('n_estimators =',best_estimators_forest,', max_depth =',best_depth_forest)

Оптимальные параметры, дающие максимальное значение метрики F1, равное 0.5513698630136986
n_estimators = 11 , max_depth = 12
CPU times: user 14.8 s, sys: 21.7 ms, total: 14.8 s
Wall time: 14.8 s


Необходимая нам метрика почти не выросла.

Попробуем третий изученный в курсе метод - метод логистической регрессии.

In [14]:
model_lr = LogisticRegression(random_state=rs, solver='liblinear')
model_lr.fit(features_train, target_train)
predicted_lr_valid = model_lr.predict(features_valid)
print('Значение F1-метрики для модели', f1_score(target_valid, predicted_lr_valid))

Значение F1-метрики для модели 0.2803347280334728


Тут значение F1-метрики совсем уж низкое. 

<b>Вывод</b>

Как бы мы не перебирали модели и их гиперпараметры, достичь необходимого значения метрики в 0,59 у нас не получается. Возможно, это связано с дисбалансом классов в исходных данных.

Попробуем несколько методов борьбы с дисбалансом, и проанализируем, какой окажется эффективнее.

___
## Борьба с дисбалансом

### Балансирование классов
<sup>[Перейти к содержанию](#Содержание)</sup>

Для борьбы с дисбалансом классов добавим в каждую обучаемую модель параметр class_weight='balanced'. Для начала, как и в прошлом разделе, рассмотрим модель древа решений.

In [15]:
%%time

max_f1 = 0
best_depth_b = 0
for depth in range(2, 16):
    model_tree_b = DecisionTreeClassifier(random_state=rs, max_depth=depth, class_weight='balanced')
    model_tree_b.fit(features_train, target_train)
    predicted_tree_valid_b = model_tree_b.predict(features_valid)
    if f1_score(target_valid, predicted_tree_valid_b) > max_f1:
        max_f1 = f1_score(target_valid, predicted_tree_valid_b)
        best_depth_b = depth

print('Оптимальная глубина дерева, дающая максимальное значение метрики F1, равное', max_f1)
print('max_depth =',best_depth_b)

Оптимальная глубина дерева, дающая максимальное значение метрики F1, равное 0.5762273901808785
max_depth = 5
CPU times: user 270 ms, sys: 51.1 ms, total: 321 ms
Wall time: 345 ms


Размер F1-метрики на отбалансированной выборке на 2% выше, чем на обычной. Не много, но все равно прогресс. Площадь под кривой при этом практически такая же, как и была. Теперь так же попробуем отбалансированные классы на случайном лесу.

In [16]:
%%time

max_f1 = 0
best_estimators_forest_b = 0
best_depth_forest_b = 0
for estimators in range(1, 51, 5):
    for depth in range(2, 16, 1):
        model_forest_b = RandomForestClassifier(random_state=rs, n_estimators=estimators, max_depth=depth, class_weight='balanced')
        model_forest_b.fit(features_train, target_train)
        predicted_forest_valid_b = model_forest_b.predict(features_valid)
        if f1_score(target_valid, predicted_forest_valid_b) > max_f1:
            max_f1 = f1_score(target_valid, predicted_forest_valid_b)
            best_estimators_forest_b = estimators
            best_depth_forest_b = depth
print('Оптимальные параметры, дающие максимальное значение метрики F1, равное', max_f1)
print('n_estimators =',best_estimators_forest_b,', max_depth =',best_depth_forest_b)

Оптимальные параметры, дающие максимальное значение метрики F1, равное 0.6185819070904646
n_estimators = 26 , max_depth = 8
CPU times: user 14.6 s, sys: 36.9 ms, total: 14.6 s
Wall time: 14.6 s


Тут всё уже гораздо лучше. Требуемый показатель F1-метрики у модели достигнут, AUC-ROC тоже на довольно высоком уровне. Посмотрим, поможет ли балансирование классов методу логистической регрессии.

In [17]:
model_lr = LogisticRegression(random_state=rs, solver='liblinear', class_weight='balanced')
model_lr.fit(features_train, target_train)
predicted_lr_valid = model_lr.predict(features_valid)
print('Значение F1-метрики для модели', f1_score(target_valid, predicted_lr_valid))

Значение F1-метрики для модели 0.4975984630163305


Тут тоже пригресс. Однако значение метрики всё еще силшком низко.

Изучим другие методы борьбы с дисбалансом классов, возможно, они окажутся еще более эффективными.

___
### Увеличение выборки
<sup>[Перейти к содержанию](#Содержание)</sup>

Попробуем увеличить выборку чтобы сбалансировать наши классы. Напишем функцию, которая будет увеличивать в выборках количество необходимых классов в нужное количество раз.

In [18]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=rs)
    
    return features_upsampled, target_upsampled

Теперь мы можем увеличить размер обучающих выборок таким образом, чтобы в них был достигнут баланс классов. Как мы помним, соотношение классов в выборке у нас было примерно 80/20. Следовательно увеличивать меньшую выборку мы будем в 4 раза.

In [19]:
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

Теперь на этих выборках можно обучить те же самые модели и иосмотреть, какой результат это дало. Так же как и ранее, делаем это в цикле, подбирая оптимальные параметры модели.

In [20]:
%%time

max_f1 = 0
best_depth_u = 0
for depth in range(2, 16):
    model_tree_u = DecisionTreeClassifier(random_state=rs, max_depth=depth)
    model_tree_u.fit(features_upsampled, target_upsampled)
    predicted_tree_valid_u = model_tree_u.predict(features_valid)
    if f1_score(target_valid, predicted_tree_valid_u) > max_f1:
        max_f1 = f1_score(target_valid, predicted_tree_valid_u)
        best_depth_u = depth

print('Максимальное значение метрики F1, равное', max_f1)
print('max_depth =',best_depth_u)

Максимальное значение метрики F1, равное 0.5762273901808785
max_depth = 5
CPU times: user 313 ms, sys: 0 ns, total: 313 ms
Wall time: 327 ms


F1-метрика не дотягивает.

In [21]:
%%time

max_f1 = 0
best_estimators_forest_u = 0
best_depth_forest_u = 0
for estimators in range(1, 51, 5):
    for depth in range(2, 16, 1):
        model_forest_u = RandomForestClassifier(random_state=rs, n_estimators=estimators, max_depth=depth)
        model_forest_u.fit(features_upsampled, target_upsampled)
        predicted_forest_valid_u = model_forest_u.predict(features_valid)
        if f1_score(target_valid, predicted_forest_valid_u) > max_f1:
            max_f1 = f1_score(target_valid, predicted_forest_valid_u)
            best_estimators_forest_u = estimators
            best_depth_forest_u = depth
print('Оптимальные параметры, дающие максимальное значение метрики F1, равное', max_f1)
print('n_estimators =',best_estimators_forest_u,', max_depth =',best_depth_forest_u)

Оптимальные параметры, дающие максимальное значение метрики F1, равное 0.5984251968503936
n_estimators = 41 , max_depth = 5
CPU times: user 20.9 s, sys: 88.1 ms, total: 21 s
Wall time: 21.2 s


Тут уже необходимая метрика достигла требуемых параметров, однако значение ниже, чем при балансировании классов.

In [22]:
model_lr = LogisticRegression(random_state=rs, solver='liblinear', class_weight='balanced')
model_lr.fit(features_upsampled, target_upsampled)
predicted_lr_valid = model_lr.predict(features_valid)
print('Значение F1-метрики для модели', f1_score(target_valid, predicted_lr_valid))

Значение F1-метрики для модели 0.49663137632338783


Модель логистической регрессии я продолжаю обучать уже больше по инерции. С самого начала было видно, что она плохо справляется, и надо было бы уже не перегружать работу ею. Каких либо особых комментариев по ней я оставлять не буду, бусть будет для сравнения.

<b>Вывод</b>

Модель на выборке, сбалансированной увеличением дала результаты ниже, чем на сбалансированной параметром class_weight='balanced'. Возможно, увеличение выборки в данной задаче - "не наш клиент". Посмотрим, что даст уменьшение выборки.

___
### Уменьшение выборки
<sup>[Перейти к содержанию](#Содержание)</sup>

В этом разделе рассмотрим метод балансировки классов уменьшением выборки. Снова напишем функцию, которая будет уменьшать в выборках преобладающий клас в необходимое количество раз.

In [23]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=rs)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=rs)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=rs)
    
    return features_downsampled, target_downsampled

In [24]:
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.25)

Скормим наши уменьшенные выборки поочередно всем трём моделям.

In [25]:
%%time

max_f1 = 0
best_depth_d = 0
for depth in range(2, 16):
    model_tree_d = DecisionTreeClassifier(random_state=rs, max_depth=depth)
    model_tree_d.fit(features_downsampled, target_downsampled)
    predicted_tree_valid_d = model_tree_d.predict(features_valid)
    if f1_score(target_valid, predicted_tree_valid_d) > max_f1:
        max_f1 = f1_score(target_valid, predicted_tree_valid_d)
        best_depth_d = depth

print('Максимальное значение метрики F1, равное', max_f1)
print('max_depth =',best_depth_u)

Максимальное значение метрики F1, равное 0.5758354755784061
max_depth = 5
CPU times: user 132 ms, sys: 11 µs, total: 132 ms
Wall time: 135 ms


И снова не густо. Все надежды на лес.

In [26]:
%%time

max_f1 = 0
best_estimators_forest_d = 0
best_depth_forest_d = 0
for estimators in range(1, 51, 5):
    for depth in range(2, 16, 1):
        model_forest_d = RandomForestClassifier(random_state=rs, n_estimators=estimators, max_depth=depth)
        model_forest_d.fit(features_downsampled, target_downsampled)
        predicted_forest_valid_d = model_forest_d.predict(features_valid)
        if f1_score(target_valid, predicted_forest_valid_d) > max_f1:
            max_f1 = f1_score(target_valid, predicted_forest_valid_d)
            best_estimators_forest_d = estimators
            best_depth_forest_d = depth
print('Оптимальные параметры, дающие максимальное значение метрики F1, равное', max_f1)
print('n_estimators =',best_estimators_forest_d,', max_depth =',best_depth_forest_d)

Оптимальные параметры, дающие максимальное значение метрики F1, равное 0.5978260869565216
n_estimators = 26 , max_depth = 7
CPU times: user 8.73 s, sys: 68 ms, total: 8.8 s
Wall time: 8.82 s


И тут повторяется результат из прошлого раздела, где мы выборку увеличивали. Результат обеих метрик не плох, однако не дотягивает до результата балансирования с помощью class_weight='balanced'.

In [27]:
model_lr = LogisticRegression(random_state=rs, solver='liblinear', class_weight='balanced')
model_lr.fit(features_downsampled, target_downsampled)
predicted_lr_valid = model_lr.predict(features_valid)
print('Значение F1-метрики для модели', f1_score(target_valid, predicted_lr_valid))

Значение F1-метрики для модели 0.5034013605442176


<b>Вывод</b>

Модель на выборке, сбалансированной уменьшением дала почти такие же результаты, как и на увеличенной выборке. Попробуем скомбинировать методы.

___
### Эксперименты
<sup>[Перейти к содержанию](#Содержание)</sup>

Кроме всего прочего поэкспериментируем, например, применив два метода балнсирования сразу, или попробуем в цикле подобрать оптимальную пропорцию апскейлинга.

Скомбинируем метод балансирования классов и увеличения выборки (он дал немного более лучший результат, нежели уменьшение).

Увеличенная выборка у нас уже подготовлена в разделе 3.2.

In [28]:
%%time

max_f1 = 0
best_estimators_forest_ub = 0
best_depth_forest_ub = 0
for estimators in range(1, 51, 5):
    for depth in range(2, 16, 1):
        model_forest_ub = RandomForestClassifier(random_state=rs, n_estimators=estimators, max_depth=depth, class_weight='balanced')
        model_forest_ub.fit(features_upsampled, target_upsampled)
        predicted_forest_valid_ub = model_forest_ub.predict(features_valid)
        if f1_score(target_valid, predicted_forest_valid_ub) > max_f1:
            max_f1 = f1_score(target_valid, predicted_forest_valid_ub)
            best_estimators_forest_ub = estimators
            best_depth_forest_ub = depth
print('Оптимальные параметры, дающие максимальное значение метрики F1, равное', max_f1)
print('n_estimators =',best_estimators_forest_ub,', max_depth =',best_depth_forest_ub)

Оптимальные параметры, дающие максимальное значение метрики F1, равное 0.600229095074456
n_estimators = 41 , max_depth = 5
CPU times: user 21.4 s, sys: 116 ms, total: 21.5 s
Wall time: 21.6 s


Неплохой результат. Теперь попробуем добавить в цикл перебор множителей апскейлинга параметров.

In [29]:
%%time

max_f1 = 0
best_estimators_rr = 0
best_depth_rr = 0
best_iter = 0
for estimators in range(1, 51, 5):
    for depth in range(2, 16, 1):
        for ite in range(2, 6):
            features_uprr, target_uprr = upsample(features_train, target_train, ite)
            model_rr = RandomForestClassifier(random_state=rs, n_estimators=estimators, max_depth=depth, class_weight='balanced')
            model_rr.fit(features_uprr, target_uprr)
            predicted_rr = model_rr.predict(features_valid)
            if f1_score(target_valid, predicted_rr) > max_f1:
                max_f1 = f1_score(target_valid, predicted_rr)
                best_estimators_rr = estimators
                best_depth_rr = depth
                best_iter = ite
print('Оптимальные параметры, дающие максимальное значение метрики F1, равное', max_f1)
print('n_estimators =',best_estimators_forest_u,', max_depth =',best_depth_forest_u,', best_iter =',best_iter)

Оптимальные параметры, дающие максимальное значение метрики F1, равное 0.6035665294924554
n_estimators = 41 , max_depth = 5 , best_iter = 5
CPU times: user 1min 21s, sys: 320 ms, total: 1min 21s
Wall time: 1min 21s


Тоже хороший результат.

<b>Общий вывод по разделу:</b>

Наилучший результат во всех случаях давала модель случайного леса. Мы подобрали параметры и обучили несколько моделей, давших необходимый результат.

| Метод борьбы с дисбалансом          | Параметры модели                  | Размер F1-метрики  |
|:------------------------------------|:----------------------------------|--------------------|
| class_weight='balanced'             | n_estimators = 26 , max_depth = 8 | 0.6185819070904646 |
| Увеличение выборки (множитель 4)    | n_estimators = 41 , max_depth = 5 | 0.5984251968503936 |
| Уменьшение выборки (множитель 0.25) | n_estimators = 26 , max_depth = 7 | 0.5978260869565216 |
| Увеличение выборки (множитель 4) + class_weight='balanced' | n_estimators = 46 , max_depth = 14 | 0.6035665294924554 |
| Подбор оптимального множителя для увеличения выборки | n_estimators = 46 , max_depth = 14 , множитель = 5 | 0.6035665294924554 |

## Тестирование модели
<sup>[Перейти к содержанию](#Содержание)</sup>

Ну и наконец, проверим наши модели на тестовой выборке. Выясним, какая из моделей даст нам наилучший результат.

Проверим все пять вариантов, из таблицы в прошлом разделе.

### Балансирование классов параметром class_weight='balanced'

In [30]:
model_test_1 = RandomForestClassifier(random_state=rs, n_estimators = 26 , max_depth = 8, class_weight='balanced')
model_test_1.fit(features_train, target_train)

predicted_test_1 = model_test_1.predict(features_test)

probabilities_test_1 = model_test_1.predict_proba(features_test)
probabilities_one_test_1 = probabilities_test_1[:, 1]
auc_roc_test_1 = roc_auc_score(target_test, probabilities_one_test_1)

if f1_score(target_test, predicted_test_1) > 0.59:
    print('Данная модель подходит! Значение F1:', f1_score(target_test, predicted_test_1))
    print('Размер метрики AUC-ROC:', auc_roc_test_1)
else:
    print('Данная модель не подходит! Значение F1:', f1_score(target_test, predicted_test_1))
    print('Размер метрики AUC-ROC:', auc_roc_test_1)

Данная модель не подходит! Значение F1: 0.5854271356783921
Размер метрики AUC-ROC: 0.8447684746119917


### Увеличение выборки на множитель 4

In [31]:
features_up_test_2, target_up_test_2 = upsample(features_train, target_train, 4)
model_test_2 = RandomForestClassifier(random_state=rs, n_estimators=41, max_depth=5)
model_test_2.fit(features_up_test_2, target_up_test_2)

predicted_test_2 = model_test_2.predict(features_test)

probabilities_test_2 = model_test_2.predict_proba(features_test)
probabilities_one_test_2 = probabilities_test_2[:, 1]
auc_roc_test_2 = roc_auc_score(target_test, probabilities_one_test_2)

if f1_score(target_test, predicted_test_2) > 0.59:
    print('Данная модель подходит! Значение F1:', f1_score(target_test, predicted_test_2))
    print('Размер метрики AUC-ROC:', auc_roc_test_2)
else:
    print('Данная модель не подходит! Значение F1:', f1_score(target_test, predicted_test_2))
    print('Размер метрики AUC-ROC:', auc_roc_test_2)

Данная модель не подходит! Значение F1: 0.5649717514124294
Размер метрики AUC-ROC: 0.8427164717216878


### Уменьшение выборки на множитель 0.25

In [32]:
features_dn_test_3, target_dn_test_3 = downsample(features_train, target_train, 0.25)
model_test_3 = RandomForestClassifier(random_state=rs, n_estimators=41, max_depth=5)
model_test_3.fit(features_dn_test_3, target_dn_test_3)

predicted_test_3 = model_test_3.predict(features_test)

probabilities_test_3 = model_test_3.predict_proba(features_test)
probabilities_one_test_3 = probabilities_test_3[:, 1]
auc_roc_test_3 = roc_auc_score(target_test, probabilities_one_test_3)

if f1_score(target_test, predicted_test_3) > 0.59:
    print('Данная модель подходит! Значение F1:', f1_score(target_test, predicted_test_3))
    print('Размер метрики AUC-ROC:', auc_roc_test_3)
else:
    print('Данная модель не подходит! Значение F1:', f1_score(target_test, predicted_test_3))
    print('Размер метрики AUC-ROC:', auc_roc_test_3)

Данная модель не подходит! Значение F1: 0.5608108108108107
Размер метрики AUC-ROC: 0.839485578587665


### Увеличение выборки на множитель 4 и параметр class_weight='balanced'

In [33]:
features_up_test_4, target_up_test_4 = upsample(features_train, target_train, 4)
model_test_4 = RandomForestClassifier(random_state=rs, n_estimators=41, max_depth=5, class_weight='balanced')
model_test_4.fit(features_up_test_4, target_up_test_4)

predicted_test_4 = model_test_4.predict(features_test)

probabilities_test_4 = model_test_4.predict_proba(features_test)
probabilities_one_test_4 = probabilities_test_4[:, 1]
auc_roc_test_4 = roc_auc_score(target_test, probabilities_one_test_4)

if f1_score(target_test, predicted_test_4) > 0.59:
    print('Данная модель подходит! Значение F1:', f1_score(target_test, predicted_test_4))
    print('Размер метрики AUC-ROC:', auc_roc_test_4)
else:
    print('Данная модель не подходит! Значение F1:', f1_score(target_test, predicted_test_4))
    print('Размер метрики AUC-ROC:', auc_roc_test_4)

Данная модель не подходит! Значение F1: 0.5671641791044776
Размер метрики AUC-ROC: 0.8422102921916632


### Подобранный оптимальный множитель для увеличения выборки

In [34]:
features_up_test_5, target_up_test_5 = upsample(features_train, target_train, 5)
model_test_5 = RandomForestClassifier(random_state=rs, n_estimators=46, max_depth=14)
model_test_5.fit(features_up_test_5, target_up_test_5)

predicted_test_5 = model_test_5.predict(features_test)

probabilities_test_5 = model_test_5.predict_proba(features_test)
probabilities_one_test_5 = probabilities_test_5[:, 1]
auc_roc_test_5 = roc_auc_score(target_test, probabilities_one_test_5)

if f1_score(target_test, predicted_test_5) > 0.59:
    print('Данная модель подходит! Значение F1:', f1_score(target_test, predicted_test_5))
    print('Размер метрики AUC-ROC:', auc_roc_test_5)
else:
    print('Данная модель не подходит! Значение F1:', f1_score(target_test, predicted_test_5))
    print('Размер метрики AUC-ROC:', auc_roc_test_5)

Данная модель подходит! Значение F1: 0.5949720670391061
Размер метрики AUC-ROC: 0.8396088304806636


___
## Общий вывод по работе
<sup>[Перейти к содержанию](#Содержание)</sup>

В итоге, мы обучили модель с необходимым значеним метрики F1 более 0,59.

Это модель случайного леса, обученная на увеличенной выборке. Подбор множителя увеличения и гиперпараметров модели были выполнены в циклах.

Полученную модель можно использовать для анализа того, уйдет ли клиент из банка в ближайшее время.

Размер метрики AUC-ROC (площадь под ROC-кривой) у всех тестируемых моделей не зависел от размера F1, и был достаточно высок, в районе 0,85. Это говорит о том, что все модели работали эффективнее случайной модели.

In [35]:
features_train_valid = pd.concat([features_train] + [features_valid])
target_train_valid = pd.concat([target_train] + [target_valid])
    
features_tv, target_tv = shuffle(
    features_train_valid, target_train_valid, random_state=rs)

In [36]:
model_test_1 = RandomForestClassifier(random_state=rs, n_estimators = 26 , max_depth = 8, class_weight='balanced')
model_test_1.fit(features_tv, target_tv)

predicted_test_1 = model_test_1.predict(features_test)
print('Значение F1:', f1_score(target_test, predicted_test_1))

Значение F1: 0.5995085995085995


---